lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitExpr;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitStmt;
+
+  void runOnOperation() override;
+
+  // Lowering module block arguments.
+  void lowerArg(BlockArgument arg, BundleType type);
+  void lowerArg(BlockArgument arg, Type type);
+
+  // Lowering operations.
+  LogicalResult visitDecl(InstanceOp op);
+  LogicalResult visitExpr(SubfieldOp op);
+  LogicalResult visitStmt(ConnectOp op);
+
+  // Skip over any other operations.
+  LogicalResult visitUnhandledOp(Operation *op) { return success(); }
+  LogicalResult visitInvalidOp(Operation *op) { return success(); }
+
+  // Helpers to manage state.
+  Value addArgument(Type type, unsigned oldArgNumber,
+                    StringRef nameSuffix = "");
+
+  void setBundleLowering(Value oldValue, StringRef flatField, Value newValue);
+  Optional<Value> getBundleLowering(Value oldValue, StringRef flatField);
+  void getAllBundleLowerings(Value oldValue, SmallVectorImpl<Value> &results);
+
+  void setInstanceLowering(Value oldValue, Value newValue);
+  Value getInstanceLowering(Value oldValue);
+
+  void setSubfieldLowering(Value subfield, Value rootBundle,
+                           std::string suffix);
+  Optional<ValueField> getSubfieldLowering(Value subfield);
+
+  void removeArg(unsigned argNumber);
+  void removeOp(Operation *op);
+  void cleanup();
+
+private:
+  // The builder is set and maintained in the main loop.
+  OpBuilder *builder;
+
+  // State to keep track of arguments and operations to clean up at the end.
+  SmallVector<unsigned, 8> argsToRemove;
+  SmallVector<Operation *, 16> opsToRemove;
+
+  // State to keep a mapping from original bundle values to flattened names and
+  // flattened values.
+  circt::DenseMap<Value, llvm::StringMap<Value>> loweredBundleValues;
+
+  // State to keep a mapping from original instances to new instances with their
+  // bundles flattened.
+  circt::DenseMap<Value, Value> loweredInstances;
+
+  // State to keep a necessary info when traversing potentially nested subfields
+  // of bundles: the root bundle and the fieldname suffix computed so far.
+  circt::DenseMap<Value, ValueField> loweredSubfieldInfo;
+};
+} // end anonymous namespace
+
+/// This is the pass constructor.
+std::unique_ptr<mlir::Pass> circt::firrtl::createLowerFIRRTLTypesPass() {
+  return std::make_unique<FIRRTLTypesLowering>();
+}
+
+// This is the main entrypoint for the lowering pass.
+void FIRRTLTypesLowering::runOnOperation() {
+  auto module = getOperation();
+  auto *body = module.getBodyBlock();
+
+  OpBuilder theBuilder(&getContext());
+  builder = &theBuilder;
+
+  // Lower the module block arguments.
+  SmallVector<BlockArgument, 8> args(body->args_begin(), body->args_end());
+  for (auto arg : args)
+    mlir::TypeSwitch<Type>(arg.getType())
+        .Case<BundleType>([&](BundleType type) { lowerArg(arg, type); })
+        .Default<>([&](Type type) { lowerArg(arg, type); });
```

Comment:
Is this typeswitch necessary?  I think the bundletype case is correct for non-bundles since flattened types turn into their own element.

Reply:
If this is needed, please use `dyn_cast` for a single test like this, it is more idiomatic.

Reply:
You're right, I initially imagined a big `TypeSwitch` handling the lowering of all sorts of FIRRTL types. But since I'm really just tackling bundletype right now, this can be simplified.